### PR TITLE
Store application metadata inside the cache entry

### PR DESCRIPTION
**Pull request by lucasuyezu**
## Problem

Applications sometimes need per-page data that must stay consistent with a cached response. Today the only option is a second cache entry keyed off `cacheable.key`, and two entries cannot be kept consistent:

- the second entry is written after the page (or, with an async store, after the response), so a hit can find the page before its companion data exists;
- it lives in a different slab class and is evicted independently;
- during stale-while-revalidate, the losers are served the *old* body while computing the *new* `cacheable.key`, so they look up companion data that does not exist yet — or, if the second write is synchronous, data that belongs to a body they were not served.

The fix is to carry that data inside the entry, next to the body it describes. The entry already has a `metadata` slot (used by the Brotli splice); this PR opens it to applications.

## Change

- `env['cacheable.metadata']` (a Hash), set by the application before the response is stored, is nested under an `app` key in the entry's metadata by `CacheWriter#cache_data`. Because both `Middleware#call` and `DeferredStore#complete` write through `CacheWriter.store`, one hook covers both paths. A non-Hash value is logged and left out; it never prevents the response from being cached.
- `ResponseCacheHandler#serve_from_cache` sets the same env key from the entry it actually serves — fresh or stale — after the response has been prepared, so a request that takes the fill lock, or falls back to a refill because decompression failed, never inherits the entry's metadata. The slot always reflects the served entry: removed when the entry carries none.
- Constants `ResponseBank::METADATA_ENV_KEY` and `ResponseBank::APP_METADATA_KEY`; README section.

## Compatibility

- Requests that do not set the env key produce byte-identical payloads (5 elements, or 6 with the existing Brotli-splice hash). Cost is one `Hash#[]` per write and one nil-safe `key?` per hit.
- The in-request `Stored#metadata` used for Brotli splice replacement is untouched; only the persisted array is extended.
- Older readers ignore the `app` key (`metadata_slots` only reads `brotli_splice`), so rolling deploys and rollbacks are safe in both directions.
- No wire-format version bump: the slot and the 6-element shape already exist since #103.

## Tests

`cache_writer_test`: nesting; unchanged shape without metadata; merge alongside a real Brotli-splice slot; non-Hash ignored with a log line and the page still stored. `response_cache_handler_test`: fresh hit exposes; entry without app metadata clears the slot, including a value set earlier in the request; stale-while-revalidate hit exposes the **stale** entry's metadata; lock-acquired miss does not expose; decompression failure falls back to a refill without publishing the rejected entry's metadata. `deferred_store_test`: `complete` picks up metadata set during rendering. `middleware_test`: end-to-end. 106 runs / 561 assertions on `gemfiles/actionpack_8.1.gemfile`, Ruby 4.0.6.

### CODE REVIEW

**Comment by lucasuyezu** [comment]
@krzksz would you be the right reviewer? This opens the metadata slot you introduced with the Brotli splice to applications, so per-page data can travel inside the entry (context in the description). Forks cannot request reviewers, hence the mention.

File: lib/response_bank/response_cache_handler.rb [hunk 1]
```diff
@@ -159,6 +159,10 @@ def serve_from_cache(cache_key_hash, match_entity_tag = "*", cache_age_tolerance
         # version check
         # unversioned but tolerance threshold
         # regen
+        if metadata&.key?(ResponseBank::APP_METADATA_KEY)
+          @env[ResponseBank::METADATA_ENV_KEY] = metadata[ResponseBank::APP_METADATA_KEY]
```
Comment (angeloocana) [inline lib/response_bank/response_cache_handler.rb]: **[P2] Avoid carrying rejected cache metadata into a refill.** This assignment happens before decompression and Brotli replacement, both of which can raise. `read_from_cache` then rescues and renders a miss, but the rejected entry's metadata remains in the env. If the renderer leaves it untouched or merges into it, the writer stores old metadata alongside the newly rendered body.

I reproduced this through the real middleware with an `ActiveSupport::Cache::MemoryStore`: an entry with invalid Brotli data and application metadata falls back to rendering, and the replacement cache entry contains the new body plus the old metadata.

Could we publish the metadata only after successful response preparation, or restore the previous env state on fallback? Extending the existing decompression-fallback test with application metadata would cover this.

File: test/cache_writer_test.rb [hunk 2]
```diff
@@ -128,6 +128,75 @@ def test_store_keeps_only_cacheable_headers
 
   private
 
+  def test_store_nests_application_metadata_in_the_entry
```
Comment (angeloocana) [inline test/cache_writer_test.rb:129]: **[P2] These four new tests are below `private`, so Minitest does not discover them.** That silently excludes the checks for application metadata nesting, unchanged entry shape, Brotli-splice coexistence, and non-Hash metadata. Could we move all four above `private` and rerun, confirming they appear in the executed test list?

File: lib/response_bank/response_cache_handler.rb [hunk 3]
```diff
@@ -159,6 +159,10 @@ def serve_from_cache(cache_key_hash, match_entity_tag = "*", cache_age_tolerance
         # version check
         # unversioned but tolerance threshold
         # regen
+        if metadata&.key?(ResponseBank::APP_METADATA_KEY)
```
Comment (angeloocana) [inline lib/response_bank/response_cache_handler.rb]: **API edge case: clear the env slot when the served entry has no application metadata.** If an application initializes `env['cacheable.metadata']` before calling the handler and then hits an older entry without `app` metadata, this conditional leaves the request's metadata intact. The caller can then mistake it for metadata belonging to the cached body, contrary to the README's promise that the key is absent for such entries.

I reproduced a successful metadata-less server hit retaining a prepopulated env Hash. Could successful hits replace or delete the slot according to the served entry, with a test that starts with the env slot already populated? This is not a blocker for the current Storefront caller, which initializes metadata on misses, but matters for the gem's public contract.

**Review by angeloocana — COMMENTED** [review]
The co-location approach looks good. I found two metadata lifecycle edge cases and a test-discovery issue below.

File: lib/response_bank/response_cache_handler.rb (hunk 1 above)
Comment (lucasuyezu) [reply lib/response_bank/response_cache_handler.rb]: Right — fixed in 3c62499: the slot is now published after the Content-Encoding block, so a decompression or splice failure that falls back to a refill never carries the rejected entry's metadata. Added `test_decompression_failure_does_not_publish_the_rejected_entrys_application_metadata` (entry with `app` metadata + `decompress` raising → dynamic body, env slot absent).

File: test/cache_writer_test.rb (hunk 2 above)
Comment (lucasuyezu) [reply test/cache_writer_test.rb:129]: Good catch, and an embarrassing one — they were never running. Moved above `private` in 3c62499; the suite goes from 100 to 106 runs and `rake test TESTOPTS=-v` lists all four. One of them needed a fix once it actually ran (`store` also logs the compression time, so the strict `expects(:log)` had to become a stub plus the specific expectation).

File: lib/response_bank/response_cache_handler.rb (hunk 3 above)
Comment (lucasuyezu) [reply lib/response_bank/response_cache_handler.rb]: Agreed. In 3c62499 a successful server hit always makes the slot reflect the served entry: set when the entry has `app` metadata, deleted otherwise. Added `test_server_cache_hit_replaces_application_metadata_set_earlier_in_the_request` (pre-populated env + metadata-less entry → key absent) and reworded the README to state the contract.

**Review by angeloocana — APPROVED** [review]

**Review by angeloocana — COMMENTED** [review]
Can you ask AI to be less verbose on the doc changes and comments?

**Comment by lucasuyezu** [comment]
Trimmed the comments and the README section in the last commit — three short comments, one short doc section.

File: lib/response_bank/cache_writer.rb [hunk 4]
```diff
@@ -109,9 +109,22 @@ def cache_data(status, representation_headers, stored, env, timestamp, content_e
...
+        metadata = entry_metadata(env, stored.metadata)
+        data << metadata if metadata
         data
       end
+
+      # A non-Hash value is logged and skipped rather than failing the write.
+      def entry_metadata(env, metadata)
+        app_metadata = env[ResponseBank::METADATA_ENV_KEY]
+        return metadata if app_metadata.nil?
+        unless app_metadata.is_a?(Hash)
```
Comment (krzksz) [inline lib/response_bank/cache_writer.rb:122]: The `is_a?(Hash)` check catches a top-level string, but a Hash with a `Time` or a plain object inside gets through and `MessagePack.dump` raises in the write block. I tried it through the middleware: the page isn't cached and the live response goes out uncompressed. So "never prevents the response from being cached" doesn't hold yet.

Two options, both fine by me:

1. Pack the app Hash on its own inside a `rescue`, drop it and log on failure.
2. Drop the check and document that the Hash must be MessagePack-serializable. The existing "Failed to write to cache" log already covers bad input.

File: README.md [hunk 5]
```diff
@@ -149,6 +149,20 @@ The headers passed to `complete` describe the cached representation. They can di
...
+
+Applications can store a Hash inside a cache entry by setting
+`env['cacheable.metadata']` before the response is stored:
+
+```ruby
+env['cacheable.metadata'] = { 'variant' => 'b' }
+```
+
+On a server cache hit, `env['cacheable.metadata']` holds the Hash stored with the
+served entry, stale or fresh; the key is removed when the entry has none. The Hash
```
Comment (krzksz) [inline README.md:162]: Worth one line saying the Hash is shared with every request served from the entry, so nothing visitor-specific should go in it. `DeferredStore#complete` has the same warning for body and headers.

**Review by krzksz — COMMENTED** [review]
Thanks for the thorough writeup, it made this easy to follow. The approach makes sense to me and I like that entries without metadata stay byte-identical.

Two small things inline. The only one I'd like resolved before merging is the Hash guard.

File: lib/response_bank/cache_writer.rb (hunk 4 above)
Comment (lucasuyezu) [reply lib/response_bank/cache_writer.rb:122]: Went with option 1 in 26ffb90: the app Hash is packed on its own first and dropped with a log line if that raises. Test added with a nested `Object.new`.

File: README.md (hunk 5 above)
Comment (lucasuyezu) [reply README.md:162]: Added in 26ffb90.

**Comment by lucasuyezu** [comment]
@krzksz comments addressed. It's ready for another review.

**Review by krzksz — APPROVED** [review]
Looks good, thanks for the quick turnaround. Verified the nested-object case is handled now and the README note covers the sharing caveat.

**Review by francispelland — APPROVED** [review]

File: lib/response_bank/cache_writer.rb [hunk 6]
```diff
@@ -157,9 +157,30 @@ def cache_data(status, representation_headers, stored, env, timestamp, content_e
...
+          ResponseBank.log("Ignoring #{ResponseBank::METADATA_ENV_KEY}: expected a Hash, got #{app_metadata.class}")
+          return metadata
+        end
+
+        begin
+          MessagePack.dump(app_metadata)
```
Comment (adrianna-chang-shopify) [inline lib/response_bank/cache_writer.rb]: Is it worth serializing _and_ deserializing the metadata here to validate it? `MessagePack.dump` confirms that metadata can serialize, but it doesn't guarantee that the cache reader can deserialize it (I'm thinking of edge cases like a deeply nested hash or unknown extension?) If we can serialize but the cache reader can't deserialize we risk not being able to use the cached entry, is that a concern?

**Review by adrianna-chang-shopify — APPROVED** [review]
One small question / comment but looks good otherwise.

File: lib/response_bank/cache_writer.rb (hunk 6 above)
Comment (lucasuyezu) [reply lib/response_bank/cache_writer.rb]: Good call — it's real and worse than "might not deserialize": msgpack-ruby's unpacker stops at 128 nested containers while the packer has no limit, so a deep Hash wrote an entry `MessagePack.load` rejected on every hit → refill → same entry rewritten, i.e. a permanent miss loop for that page.

Fixed in 6d2eb92: the merged metadata is round-tripped inside a one-element array, so nesting is checked at the depth the entry actually stores it — a standalone `load(dump(hash))` would miss depths 127–128, since the entry nests the Hash two containers deeper. Tests cover the drop + log line at the writer, and an end-to-end write→serve through the handler showing the entry is still a server hit.

Unknown ext types can't be caught writer-side (the writer can always read its own factory), so the README now says to keep it to plain MessagePack types.
